### ISSUE TITLE
loguru with --windows-console-mode=attach

### ISSUE THREAD

**Issue by Abdullah-Albanna** [Nuitka/Nuitka#3003]
- Nuitka version, full Python version, flavor, OS, etc. as output by *this exact* command.

  > python -m nuitka --version

    2.3.11
    Commercial: None
    Python: 3.11.9 (tags/v3.11.9:de54cf5, Apr  2 2024, 10:12:12) [MSC v.1938 64 bit (AMD64)]
    Flavor: CPython Official
    Executable: C:\Users\Abdullah Albanna\AppData\Local\Programs\Python\Python311\python.exe
    OS: Windows
    Arch: x86_64
    WindowsRelease: 10
    Version C compiler: ~\AppData\Local\Nuitka\Nuitka\Cache\DOWNLO~1\gcc\x86_64\13.2.0-16.0.6-11.0.1-msvcrt-r1\mingw64\bin\gcc.exe (gcc 13.2.0).

- How did you install Nuitka and Python

   - python is from the offical website, nuitka is from pip

- The specific PyPI names and versions
  > python -m pip list -v
    Package                   Version  Location                                                                            Installer
    ------------------------- -------- ----------------------------------------------------------------------------------- ---------
    altgraph                  0.17.4   C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    attrs                     23.2.0   C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    certifi                   2024.2.2 C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    cffi                      1.16.0   C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    charset-normalizer        3.3.2    C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    colorama                  0.4.6    C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    distlib                   0.3.8    C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    filelock                  3.15.1   C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    frozenlist                1.4.1    C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    idna                      3.7      C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    imageio                   2.34.1   C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    importlib_metadata        8.0.0    C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    loguru                    0.7.2    C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    multidict                 6.0.5    C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    Nuitka                    2.3.11   C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    numpy                     1.26.4   C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    ordered-set               4.1.0    C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    packaging                 24.0     C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    pefile                    2023.2.7 C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    pillow                    10.3.0   C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    pip                       24.0     C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    platformdirs              4.2.2    C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    pycparser                 2.22     C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    pywin32                   306      C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    pywin32-ctypes            0.2.2    C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    setuptools                70.0.0   C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    urllib3                   2.2.1    C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    yarl                      1.9.4    C:\Users\a\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    zipp                      3.19.2   C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip
    zstandard                 0.22.0   C:\Users\\AppData\Local\Programs\Python\Python311\Lib\site-packages pip

When I run my app with loguru, it logs in ascii characters, this can be fixed with colorama.init()

I used to allocate a new console for logging if for example -d is passed,looks something like so:
def allocateConsole() -> None:
    """
    Allocates a new console for logging.
    """
    win32console.AllocConsole()
    sys.stdout = open("CONOUT$", "w")
    sys.stderr = open("CONOUT$", "w")

def winLogsInit() -> None:
    """
    Initiate the console
    """
    if system == "Windows":
        if '-d' in sys.argv:
            allocateConsole()

now with the --windows-console-mode=attach in nuitka, I cannot do this?

how would I be able to do this?

**Comment by kayhayen** [comment]
What happens when you allocate the console? It giving an error would be normal if you already have a console, you need to check the error and ignore the one that says you already have a console, then your function is unnecessary.

Nuitka itself has no `allocateConsole` equivalent yet. Maybe one day we do also `allocate` as a mode, if it can be guaranteed to work. I assume you use `disable` mode before?

**Comment by Abdullah-Albanna** [comment]
@kayhayen 

Well yeah my way won't work if you just run a non gui app, it won't allocate a console if you already have one

Try with a gui app and compile it with nuitka with the window disabled

And obviously implement the loguru and look at how the ascii logs looking 

And to fix that I used the colorama.init() right before the logging starts and after the window allocation 


But using the nuitka window attach, the colorama.init() does not work

**Comment by kayhayen** [comment]
You should start by producing a minimal reproducer and not tell me how to create one.

**Comment by Abdullah-Albanna** [comment]
ok

to see what I mean by ascii logs, compile this

          import tkinter
          from loguru import logger
          import sys
          import win32console
          
          def allocateConsole() -> None:
          
              win32console.AllocConsole()
              sys.stdout = open("CONOUT$", "w")
              sys.stderr = open("CONOUT$", "w")
          
          
          tk = tkinter.Tk()
          
          allocateConsole()
          
          log_format = "<green>{time:YYYY-MM-DD HH:mm:ss}</green> - <level>{message}</level>"
          logger.remove()
          logger.add(sys.stdout, format=log_format)
          logger.debug("example log")
          
          tk.mainloop()

with

         nuitka --standalone --windows-console-mode=disable --enable-plugin=tk-inter python_file.py


now what I did to fix this is importing `colorama` and run the `init()` like so 

          import tkinter
          from loguru import logger
          import sys
          import win32console 
          from colorama import init
         
          
          def allocateConsole() -> None:
          
              win32console.AllocConsole()
              sys.stdout = open("CONOUT$", "w")
              sys.stderr = open("CONOUT$", "w")
          
          
          tk = tkinter.Tk()
          
          allocateConsole()
          
          init()
 
          log_format = "<green>{time:YYYY-MM-DD HH:mm:ss}</green> - <level>{message}</level>"
          logger.remove()
          logger.add(sys.stdout, format=log_format)
          logger.debug("example log")
          
          tk.mainloop()


and this works great, but not with with `--windows-console-mode=attach`, because I'm guessing `colorama.init()` does not detect the logs with the nuitka way of displaying the logs

**Comment by kayhayen** [comment]
The effect of `attach` is that Nuitka already does change the standard handles accordingly and if possible. In that case, you should get an exception from win32console.AllocConsole and you might need to handle that error that may or may not occur. If launched from explorer, with attach, I would be assuming that AllocConsole actually works, but with a console already attached, it cannot work, or can it.

**Comment by Abdullah-Albanna** [comment]
@kayhayen 

I do not want to open a new console for the logs, isn't that what `attach` purpose is?

All I want is the same effect as I told you 

I want the `colorama.init()` to detect the nuitka `attach` and do it's thing on it like what I told you


I like the `attach` better than opening a new console, but the ascii is the problem

**Comment by kayhayen** [comment]
Did you consider reading the docs, do they not say what it does? And also, I am not sure what sort of ascii problem you refer to. So far I have no idea what the actual outcome of your examples is, you failed to tell me that.

**Comment by Abdullah-Albanna** [comment]
@kayhayen 
>  you failed to tell me that.

I provided an example, with how to compile it
what do you mean I failed?

**Comment by kayhayen** [comment]
Ok, I failed to understand what the problem is, because so far, I see code that I would expect to not work, but you also didn't tell me in what way you think it's not working, what the error is, etc.  from reading the docs of AllocConsole it should be clear that you need to handle it not succeeding for there being a console already. You don't do that. Now what is Nuitka supposed to do about that?

**Comment by Abdullah-Albanna** [comment]
@kayhayen 

Alright I'll clear things out


before you guys implemented the `attach` option, I was using `AllocConsole`, which works fine, I configured it to whenever I pass in `-d` it would pop up a new console that you know, show logs on it, if I just show the logs normally, it would log in ascii characters like so
 
`\x1b[92m2024-07-16 11:29:18\x1b[0m - \x1b[94mexample log\x1b[0m`

This is not a problem or an error, it's just not readable because of `loguru` colored logs,I do not want it to look like that, it look ugly and unreadable, so what I did is `colorama.init()` and now it look like so:
`2024-07-16 11:29:18 - example log`


Now that you guys implemented the `attach`, I'd like to use it rather than `AllocConsole`, but the only problem is that it shows logs in ascii characters 


So in conclusion, all I need to do is execute the `colorama.init()` inside that `attach`, or something like that

**Comment by kayhayen** [comment]
So, if I compile only this line

```
print("\x1b[92m2024-07-16 11:29:18\x1b[0m - \x1b[94mexample log\x1b[0m")
```

It's colored when I use `--windows-console-mode=attach` with and without standalone mode. Can you confirm or deny this? Also then I tried this

```
print("\x1b[92m2024-07-16 11:29:18\x1b[0m - \x1b[94mexample log\x1b[0m")
import colorama
colorama.init()
print("\x1b[92m2024-07-16 11:29:18\x1b[0m - \x1b[94mexample log\x1b[0m")
```

And both lines are colored, as expected, since Python does the same. Are you implying that the init call should disable colors?

For Nuitka that identical behavior is mission accomplished. Whatever AllocConsole does, I don't really know, esp. not with winconsole wrapping it. 

I do however see that pythonw.exe allows printing to the console by default, which is interesting. I definitely thought pythonw.exe is more like "disable" and not closer to "attach". In my mind, "attach" would be unique to Nuitka but it's not at all I see now. And I do see that colorama.init indeed removes color when using it, which I find strange, and maybe it's more of a bug to which Nuitka's attach mode is not compatible yet.

Maybe, once I solve that, issues with forked processes having a hard time with handles for stdout/stderr are also solvable, so this is actually almost exciting.

**Comment by kayhayen** [comment]
So, maybe it's because it works barely nowhere. I am getting no output from cmd.exe Windows with pythonw but I am getting outputs from the git bash prompt. I have no clue how this can even be different. I will have to check Python source code, to figure out when they do it and when not.

**Comment by Abdullah-Albanna** [comment]
hmm, interesting, so I did 

      print("\x1b[92m2024-07-16 11:29:18\x1b[0m - \x1b[94mexample log\x1b[0m")
      import colorama
      colorama.init()
      print("\x1b[92m2024-07-16 11:29:18\x1b[0m - \x1b[94mexample log\x1b[0m")
      
and still shows like so with CMD:

![Captures](https://github.com/user-attachments/assets/2df305ec-7094-47d2-a082-f4c84289b487)

but it look fine with PowerShell:

![powershell](https://github.com/user-attachments/assets/a5eb2cb7-f15e-44b4-adcc-a8112fbc3a78)


can you please try with CMD

**Comment by kayhayen** [comment]
I looked at the code of Python there, and found no uses of AttackConsole nor AllocConsole, and merely uses existing and valid file handles, which somehow git bash manages to provide, possibly because that itself might be a GUI app, since it does its own terminal, and then it just works. And indeed, there is a difference in how `colorama.init()` behaves on the real console and on the pseudo one (?) of git bash.

**Comment by kayhayen** [comment]
Oh, my CMD does it fine for whatever reason, I suspect a Windows update.

**Comment by kayhayen** [comment]
I forgot how, but it used to be that you need to enable ANSI colors in CMD.exe manually, and I was thinking that colorama does it, but I didn't see the code for that anymore.

> ANSI colors are available by default in Windows version 1909 or newer.

Are you using that old of a build there in a VM or what. Seems colorama doesn't support that anymore. Or maybe it's actually Windows Terminal that I get if I launch CMD.exe these days.

**Comment by Abdullah-Albanna** [comment]
> there is a difference in how colorama.init() behaves on the real console and on the pseudo one (?) of git bash.

Ahh, so it's not possible to get `colorama.init()` to behave on the pseudo terminal?

> Oh, my CMD does it fine for whatever reason, I suspect a Windows update.

Are you using the cmd.exe or the microsoft terminal ?

**Comment by Abdullah-Albanna** [comment]
> ANSI colors are available by default in Windows version 1909 or newer.

Hmm, mine is 19045

if it's not supported yet then what does `colorama.init()` does to get it to work?

**Comment by kayhayen** [comment]
Build numbers are not years, the 1909 refers to their now unsupported release in 2019, so that's not it. Also one thing I am doing is to use `--run` but Nuitka enabled colors on the terminal at the time already, so the first line is colored already, running it separately makes the first line ugly indeed as it should be.

**Comment by Abdullah-Albanna** [comment]
>  Also one thing I am doing is to use --run

Yeah why is that?, when I pass in the `--run` when compiling, it shows the colors fine!

**Comment by kayhayen** [comment]
Nuitka enables ANSI colors for itself, and then there is nothing to do  for the ran program.

**Comment by Abdullah-Albanna** [comment]
is it possible to enable ANSI when using `attach` ?, like baked in the binary

**Comment by kayhayen** [comment]
So this colorama code decides different.

```
    def _winapi_test(handle):
        csbi = CONSOLE_SCREEN_BUFFER_INFO()
        success = _GetConsoleScreenBufferInfo(
            handle, byref(csbi))
        return bool(success)

    def winapi_test():
        return any(_winapi_test(h) for h in
                   (_GetStdHandle(STDOUT), _GetStdHandle(STDERR)))
```

I need to read up on why `_GetConsoleScreenBufferInfo` might be rejected, Nuitka is setting the standard handles correctly, or si I am thinking.

**Comment by kayhayen** [comment]
So, it is getting a permission denied error, when it tries to get this information which it only uses to test if it's a console handle, and if I fake it, the colors come through, but I have to assume that then redirections will have troubles, so that is not the real solution. I have to wonder, why it's not using `GetConsoleMode` which I would expect to work.

**Comment by kayhayen** [comment]
Ok, `GetConsoleMode` gives a `PermissionError` too. With the Attach, we gain access to the console, however we do not have anything better than opening `CONIO` to get a file handle, and that's not console IO and as such, is problematic, the file handle doesn't expose the console it seems.

**Comment by Abdullah-Albanna** [comment]
what have you discovered?, were you able to get it to work?

**Comment by kayhayen** [comment]
A workaround is definitely possible. The way they try to detect if it's a console doesn't work, but we will be able to add a special check for Nuitka console file handles, their name being "CONOUT$." However, it's not the time for me right now to do it properly, but it will happen eventually.

**Comment by Abdullah-Albanna** [comment]
@kayhayen 

Okay thanks, I'll use an okay solution for me until you implement it

**Comment by kayhayen** [comment]
For 4.1 the attaching got improved and the reproducer now works. The change is on the current develop branch.

**Comment by kayhayen** [comment]
Part of the 4.1 release I just made.
